library/src/G4CMPProcessUtils.cc

```diff
@@ -868,3 +880,22 @@ G4CMPProcessUtils::CreateChargeCarrier(G4int charge, G4int valley,
 
   return sec;
 }
+
+G4ThreeVector G4CMPProcessUtils::ValidateSecondaryPosition(const G4ThreeVector& pos) const {
+  G4ThreeVector secPos(pos);
+
+  // If the step is near a boundary, create the secondary in the initial volume
+  G4Navigator* nav = G4TransportationManager::GetTransportationManager()->GetNavigatorForTracking();
+  // Safety is the distance to a boundary
+  G4double safety = nav->ComputeSafety(pos);
+  // Tolerance is the error in deciding which volume a track is in
+  G4double kCarTolerance = G4GeometryTolerance::GetInstance()->GetSurfaceTolerance();
+  // If the distance to an edge is within error, we might accidentally get placed
+  // in the next volume. Instead, let's scoot a bit away from the edge.
+  if (safety <= kCarTolerance) {
+    G4ThreeVector norm = currentTrack->GetVolume()->GetLogicalVolume()->GetSolid()->SurfaceNormal(pos);
```

Comment:
The "norm" computed here should be rotated into global coordinates, for comparison with "pos."  Just do
`RotateToGlobalDirection(norm);`
which modifies 'norm' in situ.